SYS-1524: Add CLICC event signage

ztucker4 opened the pull request:
Implements [SYS-1524](https://uclalibrary.atlassian.net/browse/SYS-1524)

This PR adds a new view and template for CLICC signage. Currently, this is limited to the four CLICC classroom locations, whose `location_id` values are hard-coded in the view `display_CLICC_events()` and in the template `display_events.html`.

Restart is required before testing due to changes in environment variables and requirements. The sign itself is visible at 127.0.0.1:8000/display_CLICC_events/. There is no link from the rest of the app to this new page.

Tests have been added for the new functions in `views_utils.py`, other than `get_location_events()` which simply retrieves LibCal event data for a given location. I am still working to set up test spaces in LibCal to demonstrate potential issues with short events.

[SYS-1524]: https://uclalibrary.atlassian.net/browse/SYS-1524?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

akohler commented on signs/urls.py:
```diff
@@ -9,6 +9,9 @@
         name="display_hours",
     ),
+    path(
+        "display_CLICC_events/", views.display_CLICC_events, name="display_CLICC_events"
```
As with `get_hours_url`, please use lower-case consistently in view names and URLs.

akohler commented on signs/views.py:
```diff
@@ -42,7 +45,7 @@ def display_hours(
     """Display hours for a location. This view is used by the digital signage system."""
 
-    widget_url = settings.LIBCAL_HOURS_WIDGET
+    hours_widget_url = settings.LIBCAL_HOURS_WIDGET
```
This change needs to be made to the `get_hours()` call which follows.

akohler commented on signs/views.py:
```diff
@@ -71,6 +74,23 @@ def display_hours(
     return render(request, "signs/display.html", context)
 
 
+def display_CLICC_events(request: HttpRequest) -> HttpResponse:
```
Please use lower-case `clicc` in method names.

akohler commented on signs/views_utils.py:
```diff
@@ -121,3 +122,77 @@ def construct_display_url(
+
+def get_location_events(widget_url: str, location_id: int) -> HttpResponse:
+    """Get events for a location from the LibCal widget."""
+    widget_url = (
```
I think this is now available as `settings.LIBCAL_EVENTS_WIDGET`.  And does not need to be in `()`.

akohler commented on signs/views_utils.py:
```diff
@@ -121,3 +122,77 @@ def construct_display_url(
+        start, end = event["times"].split(" - ")
+        end = end.split(" ")[0]
+
+        start_time = datetime.strptime(start, "%I:%M%p").time()
```
The same format, `%I:%M%p`, is used several times.  It's more meaningful and easier to maintain to assign this to a variable (documented with comment), and reference that variable in all of these `strptime()`s.

akohler commented on signs/views_utils.py:
```diff
@@ -121,3 +122,77 @@ def construct_display_url(
+    # 8am is row 2, 8:30am is row 3, 9am is row 4, etc.
+    hour = time.hour
+    minute = time.minute
+    # round down to the nearest half hour, for simplicity and consistency
+    if minute < 30:
+        minute = 0
+    else:
+        minute = 30
+    # if time is on the half hour, add 0.5 to the hour so it will
+    # be one row lower
+    if minute == 30:
+        hour += 0.5
```
I think this can be simpler, since ultimately we don't care what `minute` is.  So perhaps this, with comments reworked to reflect the code change:
```
if minute >= 30:
    hour += 0.5
```

akohler commented on static/css/events.css:
```diff
@@ -0,0 +1,191 @@
+  color: white;
+  background-repeat: no-repeat;
+  background-size: 1080px 1920px;
+  background-image: url("/static/portrait_background_small.png");
```
File does not exist; should be `./static/img/portrait_background_small.png`

akohler commented on signs/views_utils.py:
```diff
@@ -121,3 +122,77 @@ def construct_display_url(
+    widget_url = (
+        "https://calendar.library.ucla.edu/api_events.php?m=today&simple=ul_date&cid="
+    )
+    events = []
```
`Flake8` says `events` is never used.

akohler reviewed (changes requested):
@ztucker4 Quite a few small changes are needed, caught with `Flake8` - please make sure that's installed locally.  Also some refactoring suggestions, especially if a small change to the data passed to the events template could eliminate repeated code.

A bigger problem is the way the output looks for me - see screenshot below.  This could be related to the missing background image in `events.css`.  It may also be related to my Windows use of a `Dark` theme, which caused problems with the CSS initially used with the LBS app.  However, I just switched to the default `Light` theme and it didn't help, so we'll need to see if this is a "works on my computer" situation.  Working on the signs is most important, but could be confusing for librarians if it's inconsistent.

Here's what I see:
![sign_events_hours_problem](https://github.com/UCLALibrary/digital-signage-hours/assets/2213836/c69bcacb-8246-44aa-bf5b-a87625e92da0)

ztucker4 commented:
Hi @akohler, thank you for all the feedback! I've fixed my Flake8 installation and implemented the suggested fixes, so this is ready for another look.

akohler commented on signs/templates/signs/display_events.html:
```diff
@@ -0,0 +1,53 @@
+            <div class="time-marker six-pm">6 PM</div>
+            <div class="time-marker seven-pm">7 PM</div>
+            <div class="time-marker eight-pm">8 PM</div>
+            {% for location_id, events in location_events.items %}
```
Beautiful compact code! 💯

akohler commented on signs/views.py:
```diff
@@ -71,6 +74,27 @@ def display_hours(
+
+    events_widget_url = settings.LIBCAL_EVENTS_WIDGET
+    # location IDs for CLICC classrooms, with corresponding names used as CSS classes
+    locations = [(3363, "classA"), (4357, "classB"), (4358, "classC"), (4799, "inq3")]
```
Since these values have meanings, a dict is a better structure than a tuple.  Then the `location[0]` and `location[1]` below can be replaced with `location["location_id"]` etc.

It is more verbose, but self-documenting code is generally better.

akohler reviewed (changes requested):
@ztucker4 This is much cleaner, thank you for making the changes.  The events now display perfectly for me.

Just one more refinement to your updated code, switching from a tuple to a dict, then this will be merged.

akohler commented:
@ztucker4 All is great now, merging.
